planner/core/stats.go

```diff
@@ -178,19 +179,51 @@ func (ds *DataSource) DeriveStats(childStats []*property.StatsInfo) (*property.S
 		}
 	}
 	// Consider the IndexMergePath. Now, we just generate `IndexMergePath` in DNF case.
-	if len(ds.pushedDownConds) > 0 && len(ds.possibleAccessPaths) > 1 && ds.ctx.GetSessionVars().GetEnableIndexMerge() {
-		needConsiderIndexMerge := true
-		for i := 1; i < len(ds.possibleAccessPaths); i++ {
-			if len(ds.possibleAccessPaths[i].accessConds) != 0 {
-				needConsiderIndexMerge = false
-				break
+	if len(ds.pushedDownConds) > 0 && len(ds.possibleAccessPaths) > 1 &&
+		ds.ctx.GetSessionVars().GetEnableIndexMerge() {
+		ds.generateAndPruneIndexMergePath()
+	} else if ds.indexMergeHints != nil {
+		ds.indexMergeHints = nil
+		ds.ctx.GetSessionVars().StmtCtx.AppendWarning(errors.Errorf("IndexMerge is inapplicable or disable"))
+	}
+
+	return ds.stats, nil
+}
+
+func (ds *DataSource) generateAndPruneIndexMergePath() {
+	regularPathCount := len(ds.possibleAccessPaths)
+	ds.generateIndexMergeOrPaths()
+	if ds.indexMergeHints != nil {
+		if regularPathCount < len(ds.possibleAccessPaths) {
+			// Do not need to consider the regular paths in find_best_task().
+			ds.possibleAccessPaths = ds.possibleAccessPaths[regularPathCount:]
+			// If index merge hint only specifies the table name, we should return the index merge
+			// path with minimal cost. if index merge hint specifies the table name and indexes, we
+			// should return the index merge path with the most covered indexes and minimal cost.
+			if ds.indexMergeHints[0].IndexNames != nil {
+				newAccessPaths := make([]*accessPath, 0, len(ds.possibleAccessPaths))
+				currentCoverdCount := 0
+				for _, path := range ds.possibleAccessPaths {
+					current := len(path.partialIndexPaths)
+					if current > currentCoverdCount {
```

Comment:
This heuristic does not make sense IMHO, why not just keeping these paths and let cost model decide which one to choose?